v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/PipelineController.java

```diff
@@ -220,88 +102,94 @@ static PipelineResult executeShardedMigration(
-          String suffix = generateSuffix(shardId, srcTable);
-          String shardIdColumn =
-              schemaMapper.getShardIdColumnName(
-                  reader.getSourceSchema().schemaReference().namespace(), srcTable);
-          PCollection<Void> output =
-              pipeline.apply(
-                  "Migrate" + suffix,
-                  new MigrateTableTransform(
-                      options, spannerConfig, ddl, schemaMapper, reader, shardId, shardIdColumn));
-          outputs.put(srcTable, output);
-        }
-        // Add transform to increment table counter
-        Map<String, Wait.OnSignal<?>> waitOnsMap =
-            outputs.entrySet().stream()
-                .collect(
-                    Collectors.toMap(Map.Entry::getKey, mapEntry -> Wait.on(mapEntry.getValue())));
-        pipeline.apply(
-            "Increment_table_counters_" + shardId, new IncrementTableCounter(waitOnsMap, shardId));
+
+        // If a namespace is configured for a shard uses that, otherwise uses the namespace
+        // configured in the options if there is one.
+        String namespace = Optional.ofNullable(shard.getNamespace()).orElse(options.getNamespace());
+
+        ShardedDbConfigContainer dbConfigContainer =
+            new ShardedDbConfigContainer(
+                shard, sqlDialect, namespace, shardId, entry.getKey(), options);
+        setupLogicalDbMigration(
+            options,
+            pipeline,
+            spannerConfig,
+            tableSelector,
+            levelToSpannerTableList,
+            dbConfigContainer);
       }
     }
     return pipeline.run();
   }
 
-  // Calculate the total number of logical shards in the list of physical shards.
-  private static long findNumLogicalshards(List<Shard> shards) {
-    return shards.stream().mapToLong(shard -> shard.getDbNameToLogicalShardIdMap().size()).sum();
+  private static void setupLogicalDbMigration(
+      SourceDbToSpannerOptions options,
+      Pipeline pipeline,
+      SpannerConfig spannerConfig,
+      TableSelector tableSelector,
+      Map<Integer, List<String>> levelToSpannerTableList,
+      DbConfigContainer configContainer) {
+    Map<Integer, PCollection<Void>> levelVsOutputMap = new HashMap<>();
+    for (int currentLevel = 0; currentLevel < levelToSpannerTableList.size(); currentLevel++) {
+      List<String> spannerTables = levelToSpannerTableList.get(currentLevel);
+      LOG.info("processing level: {} spanner tables: {}", currentLevel, spannerTables);
+      List<String> sourceTables =
+          spannerTables.stream()
+              .map(t -> tableSelector.getSchemaMapper().getSourceTableName("", t))
+              .collect(Collectors.toList());
+      LOG.info("level: {} source tables: {}", currentLevel, spannerTables);
+      PCollection<Void> previousLevelPCollection = levelVsOutputMap.get(currentLevel - 1);
+      if (currentLevel > 0 && previousLevelPCollection == null) {
+        LOG.warn(
```

Comment:
In what scenarios will `previousLevelPCollection` be null despite having parents dependency?

Reply:
If a list of tables does not include parent, then that will come up in the table selector. So i don't think this will come up, but i've added a log just in case some edge case we've not thought of crops up.